common/internal_model/src/main/scala/uk/ac/wellcome/models/parse/FreeformDateParser.scala

```diff
@@ -0,0 +1,180 @@
+package uk.ac.wellcome.models.parse
+
+import java.time.LocalDate
+
+import fastparse._, NoWhitespace._
+
+/**
+  *  Attempts to parse freeform text into a date or date range
+  */
+object FreeformDateParser extends Parser[FreeformDate] with DateParsingUtils {
+
+  def parser[_: P] =
+    Start ~ (inferredDate | knownDate) ~ End
+
+  def knownDate[_: P] =
+    (dateRange | fuzzyDate) map { FreeformDate(_) }
+
+  def inferredDate[_: P] =
+    ("[".? ~ knownDate ~ "]") map (_ withInferred true)
+
+  def fuzzyDate[_: P] =
+    FuzzyDateParser.parser map { Left(_) }
+
+  def dateRange[_: P] =
+    DateRangeParser.parser map { Right(_) }
+}
+
+/**
+  *  Attempts to parse freeform text into a date range
+  */
+object DateRangeParser extends Parser[DateRange] with DateParsingUtils {
+
+  def parser[_: P] =
```

Comment:
Why `[_: P]`?

Reply:
This is a context bound. Am following the FastParse documentation for this, without it we get the error:

```
could not find implicit value for parameter ctx: fastparse.P[_]
```

Reply:
Yup - i'm familiar with type bounds - guessed it was probably "so the compiler doesn't complain" but it's just a bit confusing.

Reply:
Could we leave a note with this so we won't have to look it up when confused later?

Reply:
As this syntax is used absolutely everywhere in the parsers, where would the best place to leavev it be? In the parser trait [here](https://github.com/wellcometrust/catalogue/pull/70/files#diff-f833ec2c3fb7d5264ebc6086d702749bR15)?